### ISSUE TITLE
Singular Resources

### ISSUE THREAD

**Issue by lindyhopchris** [cloudcreativity/laravel-json-api#10]
Add support for registering routes of singular resources:
http://discuss.jsonapi.org/t/does-jsonapi-love-singular-resources/125/5

**Comment by dmitry-php** [comment]
Hi @lindyhopchris 
I think it can be easily implemented by comment
`throw new RuntimeException('No matching resource type from the current route name.');`
in RequestInterpreter class. 

So after this we can add any route we want and place custom actions in same controller.

With this solution I added **/users/me** route and map it to method in UsersController class. This class extend EloquentController and serve all json-api stuff.

Is it good solutions? Or not? What side effects can we get with such solution?

Thanks in advance

**Comment by dmitry-php** [comment]
or maybe @jstoone have thoughts about this

**Comment by lindyhopchris** [comment]
If that works at the moment then I suppose it might be acceptable at the moment.

I think the more permanent solution going forward would be to use the options when registering the JSON-API resource with the route `JsonApi::resource('users')` to pass in something that registers `me` as an extra route on the same controller.

That should be fairly straight-forward to do, but we'd then need to follow through what happens with all the parsing of the request to check that it doesn't start throwing exceptions elsewhere. The main problem is that because the endpoint is not defined in the spec, it's not really possible to know how to validate the request content for a `POST` or `PUT` request.

**Comment by jstoone** [comment]
This sounds like a really neat feature!

The first thought that hit me, was that these singular resources sound very much like an alias to me. And since the singular resources are ideally going to be quite few, I'd actually suggest an implementation that could fit in either the `RouteServiceProvider` or route file without being to intrusive, which would allow us to achieve a resource alias api like:
```php
// The array key is the JSON:API `resourceType`
// The array value is the ID identifier for the resource
// The last string argument is the final alias to get the resource

JsonApi::resourceAlias(
    ['users' => auth()->id()]
    '/users/me'
);
```

_NB: for simplicity sake lets assume the currently authenticated user has an ID of **5**_
Then we can logically keep it simple:
- Register alias within a container
- When request arrives we can check if the resource type has an alias
  - In this case the resource type is `users`
- If we have an alias, we can transform the current aliased request to the specific path
  - In the example above we would transform from `/users/me` to `/users/5`
- Now we can continue processing as if the request was `/users/5` from the start

I might not have understood or not think the solution through, so I hope it makes sense @dmitry-php @lindyhopchris. 😄

**Comment by jstoone** [comment]
@lindyhopchris I think I'll start taking a look at this, and then we can iterate if it does not seem to fit the style you have in mind. 🙂 

All of a sudden laravel-json-api got a lot of activity. Nice! Let the work beigin! 😅

**Comment by lindyhopchris** [comment]
@jstoone sorry, meant to reply earlier. The alias won't work because you won't have the authenticated user id at the time that you are registering routes.

**Comment by lindyhopchris** [comment]
Closing this as it is effectively solved via #196 

I.e. in `1.0.0-rc.1` you'll be able to add custom actions for a resource, so you could do this to add a `me` route for the `users` resource:

```php
JsonApi::register('default')->withNamespace('Api')->routes(function ($api) {
    $api->resource('users')->controller()->routes(function ($posts) {
        $posts->get('me', 'current');
    });
});
```

Which would mean `GET /api/users/me` would be handled by the `Api\UsersController@current` action.
